### PR TITLE
Steps for submitting pull requests

### PR DESCRIPTION
**Pull request by afili**

### CODE REVIEW

**Comment by jgravois** [comment]
thanks @afili!

maybe this is overkill, but i'm thinking that it'd be helpful to get a little more elaborate with the instructions, include screenshots, and to break out the topic into its own .md file (linked from contributing)).

some things that would be relevant
1. describing exactly who can submit Pull Requests (not just contributors, but anyone)
2. linking to any external resources that are helpful for this particular topic.
3. walking through the key steps using a popular git client

i dont think we need to worry about documenting step by step instructions for every single git client out there, but if we go to that much trouble with the couple most popular (Sourcetree, Github for Windows) it could go a long way toward encouraging more folks to get involved.

anyone else have an opinion about all this?  i'm also happy to merge what we have now and call what i've outlined above a more longterm goal that someone can tackle at their leisure.

**Comment by AshleyDesktop** [comment]
This is really helpful, thanks! For those external resources:

"Make sure the local master is up to date with the commits upstream". I just created a pull request with conflicts because I forgot (/didn't know) to do this. Here was doc I found helpful.

Syncing a fork
https://help.github.com/articles/syncing-a-fork/

"(optional) Squash unique commits down into one when everyone is happy with the PR". This doc might be helpful for squashing commits:

squashing commits with rebase
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

**Comment by jgravois** [comment]
glad to hear it @AshleyDesktop!  in general, making a habit of keeping `master` clean so that you can conveniently "fast-forward" (see discussion a little-ways into the Basic Branching topic [here](http://git-scm.com/book/en/v2/Git-Branching-Basic-Branching-and-Merging#Basic-Branching) after new upstream commits have been merged is the key to success.

granted, describing the steps to accomplish this makes the whole prospect of contributing a tad more complicated in the short-term, but i'd argue its well worth it, because it makes interacting with pretty much every Github repository infinitely less headaching inducing after you've memorized the steps.

if you'd be willing to collaborate with @afili on some doc to clarify this process, it'd be insanely helpful.
in general, i think our best bet would be to walkthrough the scenario in which the contributor's pre-existing fork has 'gone stale', since it will apply to people who are creating their fork for the first time with only the slightest differences in steps.

**Comment by afili** [comment]
thank you both! I included these links in the guide.

I created a more detailed doc with instructions and screen shots from sourceTree. You can see it at my fork here https://github.com/afili/developer-support/blob/SubmitPR/SUBMITPR.md 

It has like 1000 commits now (!) but I will squash them once we decide how we want it to look and do all the necessary changes.

File: SUBMITPR.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+### Submitting a pull request
+
+Pull requests let you tell others about changes you've pushed to a repository on GitHub. Once a pull request is sent, interested parties can review the set of changes, discuss potential modifications, and even push follow-up commits if necessary. Everyone can submit pull request and they're useful in the shared repository model where they're used to initiate code review and general discussion about a set of changes before being merged into a mainline branch.
```
Comment (jgravois) [inline SUBMITPR.md]: * 'once a pull request is *created*'
* '*Anyone* can submit *a* pull request'

File: SUBMITPR.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+* Add a new code sample, and once you are satisfied with the addition, commit the change by clicking the "Commit" button in SourceTree. Before you commit the changes, you will have to provide a commit message that describes your changes.
+
+![alt text](images/commit.PNG "Commit")
+
+* Push the change to your fork. Click the "Push" button and make sure to select the correct repository and branch to push. You will want to push the changes to your fork and not to the upstream directly.
+
+![alt text](images/push.PNG "Push")
+
+* Use Github to create your Pull Request. Click the pull requests button and select "New pull request". The comparing changes page will open. You can click the button "compare across forks" and then select your fork and branch as shown in the picture below.
```
Comment (jgravois) [inline SUBMITPR.md]: * 'Click the pull requests button on the repository [homepage](https://github.com/esri/developer-support/) and select..'

**Comment by jgravois** [comment]
this is absolutely awesome @afili.  i made a few small comments, but afterward we are ready to merge.

**Comment by afili** [comment]
thanks for reviewing @jgravois . I did these changes and also added some extra text in the beginning that explains about the remotes that we need to have. Let me know if we want to change something there otherwise we can merge.

File: SUBMITPR.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
+### Submitting a pull request
+
+Pull requests let you tell others about changes you've pushed to a repository on GitHub. Once a pull request is created, interested parties can review the set of changes, discuss potential modifications, and even push follow-up commits if necessary. Anyone can submit a pull request and they're useful in the shared repository model where they're used to initiate code review and general discussion about a set of changes before being merged into a mainline branch.
+
+This guide outlines the step by step instructions for submitting a pull request using the SourceTree desktop client application (http://www.sourcetreeapp.com/) 
+
+* Before you start the process, make sure that you have set up the remotes. In order to do so, you can simply right-click on the remotes in SourceTree and select "New remote". You need to have one remote that will point to your remote fork (for example afili/developer-support) and one that will point to the original repository (esri/developer-support) as shown in the picture below
```
Comment (jgravois) [inline SUBMITPR.md]: * Before you start the process, make sure that you have set up *remote tracking.  This is what helps you keep track of changes to the fork and origin repositories that Github is hosting.* In order to do this, you can simply right-click on the remotes in SourceTree and select "New remote". *One will point to your remote fork (for example afili/developer-support) and one will point to the original repository (esri/developer-support) as shown in the picture below.  **Note:** *You can name them whatever you want, but it has become a convention to name the fork 'origin' and the original repository 'upstream', so its helpful to use those names if you plan on perusing StackOverflow for help.*

**Comment by jgravois** [comment]
i'm happy with this.  please take another stab at rebasing/squashing and let me know if you need a hand.  i have a feeling it will go much smoother this time. :smile:

**Comment by afili** [comment]
thanks @jgravois . @nohe427 and I just squashed so it should be ready to go.

**Comment by jgravois** [comment]
thanks so much for your work on this.  it'll make a great resource for future contributors and is a good example for anyone that might want to tackle #18 (or some other tutorial) as well.

**Pull request merged**